File: FORCES_client/examples/Python/HighLevelInterface/BasicExampleSoftConstraints/basic_example_soft_constraints.py
import numpy as np
import matplotlib.pyplot as plt
import forcespro
import forcespro.nlp
import casadi 

"""
Simple MPC - double integrator example for use with FORCESPRO
with soft inequality constraints on u

 min   xN'*P*xN + sum_{i=1}^{N-1} xi'*Q*xi + ui'*R*ui
xi,ui
     s.t. x1 = x
       x_i+1 = A*xi + B*ui  for i = 1...N-1
       xmin <= xi <= xmax   for i = 1...N
       umin <= ui <= umax   for i = 1...N

 and P is solution of Ricatti eqn. from LQR problem

(c) Embotech AG, Zurich, Switzerland, 2013-2021
"""


# Model Definition
# ----------------

# system
A = np.array([[1.1, 1], [0, 1]])
B = np.array([[1], [0.5]])
nx, nu = np.shape(B)
lam = 8 # measure for penalty term

# MPC setup
N = 10
Q = np.eye(nx)
R = np.eye(nu)
P = 10*Q
umin = -0.5
umax = 0.5
xmin = np.array([-5, -5])
xmax = np.array([5, 5])

# FORCESPRO multistage form
# assume variable ordering zi = [si; ui; xi] for i=1...N

# dimensions
model = forcespro.nlp.SymbolicModel(11)   # horizon length
model.nvar = 4    # number of variables
model.neq = 2   # number of equality constraints
model.nh = 2

# objective with penalty term
model.objective = (lambda z: z[1]*R*z[1] + lam*z[0] +
                   casadi.horzcat(z[2], z[3]) @ Q @ casadi.vertcat(z[2], z[3]))
model.objectiveN = (lambda z: z[1]*R*z[1] + lam*z[0] *
                    casadi.horzcat(z[2], z[3]) @ P @ casadi.vertcat(z[2], z[3]))

# equalities
model.eq = lambda z: casadi.vertcat(casadi.dot(A[0, :], casadi.vertcat(z[2], z[3])) + B[0, :]*z[1],
                               casadi.dot(A[1, :], casadi.vertcat(z[2], z[3])) + B[1, :]*z[1])
              
model.E = np.concatenate([np.zeros((2, 2)), np.eye(2)], axis=1)

# initial state
model.xinitidx = [2, 3]

# relaxed inequalities
model.ineq = lambda z: casadi.vertcat( z[1] - z[0],
                                       z[1] + z[0])
model.hu = np.array([umax, +float('inf')])
model.hl = np.array([-float('inf'), umin])

# inequalities
model.lb = np.concatenate([[0, -float('inf')], xmin])
model.ub = np.concatenate([[float('inf'), float('inf')], xmax])


# Generate FORCESPRO solver
# -------------------------

# set options
options = forcespro.CodeOptions()
options.printlevel = 0
options.overwrite = 1
options.nlp.bfgs_init = None

# generate code
solver = model.generate_solver(options)

# Run simulation
# --------------

x1 = [-4, 2]
kmax = 30
x = np.zeros((2, kmax + 1))
x[:, 0] = x1
u = np.zeros((1, kmax))
s = np.zeros((1, kmax))
problem = {}

solvetime = []
iters = []

for k in range(kmax):
    problem["xinit"] = x[:, k]

    # call the solver
    solverout, exitflag, info = solver.solve(problem)
    assert exitflag >= 0, "Some problem in solver"

    s[:, k] = solverout["x01"][0]
    u[:, k] = solverout["x01"][1]
    solvetime.append(info.solvetime)
    iters.append(info.it)

    x[:, k + 1] = model.eq(np.concatenate([s[:, k], u[:, k], x[:, k]])).full().reshape(2,)


# Plot results
# ------------

fig = plt.gcf()
plt.subplot(2,1,1)
plt.grid('both')
plt.title('states')
plt.plot([1, kmax], [5, 5], 'r--')
plt.plot([1, kmax], [-5, -5], 'r--')
plt.ylim(1.1*np.array([-5, 5]))
plt.step(range(1, kmax+1), x[0, range(1, kmax + 1)])
plt.step(range(1, kmax+1), x[1, range(1, kmax + 1)])

plt.subplot(2,1,2)
plt.grid('both')
plt.title('input')
plt.plot([1, kmax], [0.5, 0.5], 'r--')
plt.plot([1, kmax], [-0.5, -0.5], 'r--')
plt.ylim(1.1*np.array([-1, 1]))
plt.step(range(1, kmax+1), u[0, range(0, kmax)])

plt.show()

File: FORCES_client/examples/matlab/HighLevelInterface/BasicExampleSoftConstraints/BasicExampleSoftConstraints.m
% Simple MPC - double integrator example with soft constraints
% on input variable u for use with FORCESPRO
% 
%  min   xN'*P*xN + sum_{i=1}^{N-1} xi'*Q*xi + ui'*R*ui
% xi,ui
%       s.t. x1 = x
%            x_i+1 = A*xi + B*ui  for i = 1...N-1
%            xmin <= xi <= xmax   for i = 1...N
%            umin <= ui <= umax   for i = 1...N
%
% and P is solution of Ricatti eqn. from LQR problem
%
% (c) Embotech AG, Zurich, Switzerland, 2015-2021.

%% system
A = [1.1 1; 0 1];
B = [1; 0.5];
[nx,nu] = size(B);
lambda = 8; % measure for penalty term

%% MPC setup
N = 10;
Q = eye(nx);
R = eye(nu);
if( exist('dlqr','file') )
    [~,P] = dlqr(A,B,Q,R);
else
    P = 10*Q;
end
umin = -0.5;     umax = 0.5;
xmin = [-5, -5]; xmax = [5, 5];

%% FORCESPRO multistage form
% assume variable ordering zi = [si; ui; xi] for i=1...N

% dimensions
model.N     = 11;   % horizon length
model.nvar  = 4;    % number of variables
model.neq   = 2;    % number of equality constraints
model.nh    = 2;    % number of inequality constraints

% objective with penalty term
model.objective = @(z)  z(2)*R*z(2) + [z(3);z(4)]'*Q*[z(3);z(4)] + lambda*z(1);
model.objectiveN = @(z) z(2)*R*z(2) + [z(3);z(4)]'*P*[z(3);z(4)] + lambda*z(1); 

% equalities
model.eq = @(z) [ A(1,:)*[z(3);z(4)] + B(1)*z(2);
                  A(2,:)*[z(3);z(4)] + B(2)*z(2)];
              
model.E = [zeros(2,2), eye(2)];

% initial state
model.xinitidx = 3:4;

% relaxed inequalities
model.ineq = @(z) [ z(2) - z(1);
                    z(2) + z(1)];
model.hu = [umax, +inf];
model.hl = [-inf, umin];


model.lb = [0,      -inf, xmin  ];
model.ub = [+inf,   +inf, xmax  ];


%% Generate FORCESPRO solver

% get options
codeoptions = getOptions('FORCESNLPsolver');
codeoptions.printlevel = 2;
codeoptions.nlp.ad_tool = 'casadi';
%codeoptions.nlp.ad_tool = 'symbolic-math-tbx';

% generate code
FORCES_NLP(model, codeoptions);


%% simulate
x1 = [-4; 2];
kmax = 30;
X = zeros(2,kmax+1); X(:,1) = x1;
U = zeros(1,kmax);
S = zeros(1,kmax);
problem.x0 = zeros(model.N*model.nvar,1); 
for k = 1:kmax
    
    problem.xinit = X(:,k);
    
    [solverout,exitflag,info] = FORCESNLPsolver(problem);

    if( exitflag == 1 )
        S(:,k) = solverout.x01(1);
        U(:,k) = solverout.x01(2);
        solvetime(k) = info.solvetime;
        iters(k) = info.it;
    else
        error('Some problem in solver');
    end
    
    %X(:,k+1) = A*X(:,k) + B*U(:,k);
    X(:,k+1) = model.eq( [S(:,k);U(:,k);X(:,k)] )';
end

%% plot
figure; clf;

subplot(2,1,1); grid on; title('states'); hold on;
plot([1 kmax], [5 5]', 'r--'); plot([1 kmax], [-5 -5]', 'r--');
ylim(1.1*[min(-5),max(5)]); stairs(1:kmax,X(:,1:kmax)');

subplot(2,1,2);  grid on; title('input'); hold on;
plot([1 kmax], [0.5 0.5]', 'r--'); plot([1 kmax], [-0.5 -0.5]', 'r--');
ylim(1.1*[min([U, -0.5]),max([U, 0.5])]); stairs(1:kmax,U(:,1:kmax)');